### PR TITLE
dispatch frame methods to series versions instead of re-implementing masking etc

### PR DESCRIPTION
**Pull request by jbrockmendel**
This moves some of DataFrame's dispatching logic into `ops`, mostly in ways that don't change any logic.

The place that logic _is_ changed is in `DataFrame._combine_frame` where instead of defining `arith_op` to mask `func` and wrapping `{col: arith_op(this[col].values, other[col].values) for col in this.columns}`, we dispatch directly to the Series methods and wrap `{col: func(this[col], other[col]) for col in this.columns}`

### CODE REVIEW

File: pandas/core/frame.py
Comment (jbrockmendel) [inline pandas/core/frame.py]: I think we can do better than operating on `left.values.T` and `right.values`.  Especially with mixed-dtypes this seems like low-hanging fruit.

File: pandas/core/frame.py
Comment (jbrockmendel) [inline pandas/core/frame.py]: I'd like to dispatch to Series here, but presumably doing the operation once instead of per-column has a perf benefit.  One idea that didn't work on the first try was to ravel() `this.values` and `other.values`, wrap them in `Series`, then operate and re-shape.

File: pandas/core/frame.py
Comment (jreback) [reply pandas/core/frame.py]: most of these types of operations need to be pushed down to internals via block operations. this should dispatch to ._data.eval to do this (like the other match functions)

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
> The place that logic is changed is in DataFrame._combine_frame where instead of defining arith_op to mask func and wrapping {col: arith_op(this[col].values, other[col].values) for col in this.columns}, we dispatch directly to the Series methods and wrap {col: func(this[col], other[col]) for col in this.columns}

does this change anything from a user pov?

File: pandas/core/frame.py
Comment (jbrockmendel) [reply pandas/core/frame.py]: Yah, but I'm not there yet.  These particular cases are tricky because SparseDataFrame implements `_combine_match_foo` differently

**Comment by jbrockmendel** [comment]
> does this change anything from a user pov?

No, but I may have made a mistake here.  Pls put a pin in this until I double-check.

**Comment by jbrockmendel** [comment]
I did in fact make a mistake, but there are no existing tests that catch it.  Next push will fix both of these.

**Comment by jbrockmendel** [comment]
Will hold off on this until #19613 is closed.

File: pandas/core/ops.py
Comment (jreback) [inline pandas/core/ops.py]: can you add a doc-string

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
rebase

**Comment by jreback** [comment]
did you have another PR which touched this code and we wanted to merge first? or is this one ok? if so rebase

**Comment by jbrockmendel** [comment]
There was, and it was merged a few days ago.  #19613.  This should be good to go (after rebase)

**Comment by jbrockmendel** [comment]
#19649 touches _nearby_ code, but distinct enough that it shouldn't cause a problem.

File: pandas/core/ops.py
Comment (jreback) [inline pandas/core/ops.py]: you might need to parameterize this on the work 'index' but not sure (IOW can it be a Series)? or maybe just remove the word index ?

File: pandas/core/ops.py
Comment (jreback) [inline pandas/core/ops.py]: use the format
``fill_value : object, default None``

no parens

File: pandas/core/ops.py
Comment (jreback) [inline pandas/core/ops.py]: we normally use axis numbers internally (and not names). I agree its a bit easier on the eyes to do this, but we should be consistent (so revert here, ok to change it all over the base in another PR).

File: pandas/tests/frame/test_arithmetic.py
Comment (jreback) [inline pandas/tests/frame/test_arithmetic.py]: does this represent a change in user facing API?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/ops.py
Comment (jbrockmendel) [reply pandas/core/ops.py]: ATM it is only used for indexes.  I could add an assertion that self is an Index so that we know to change it if/when the wording ceases to be accurate.

File: pandas/core/ops.py
Comment (jbrockmendel) [reply pandas/core/ops.py]: OK.  This isn't a change from the current implementation, just moving it from frame.  Will change.

File: pandas/tests/frame/test_arithmetic.py
Comment (jbrockmendel) [reply pandas/tests/frame/test_arithmetic.py]: No, just a case that is not currently tested.  I made a mistake in an early commit in this PR; this test would have caught it.

File: pandas/core/frame.py
Comment (jreback) [reply pandas/core/frame.py]: this is pretty tricky, yes that would work for most ufuncs, but is not generally applicable e.g. concat is a ufunc too.
so hold off on this as other low hanging fruit w/o opening up this box.

Comment (jreback) [reply pandas/core/frame.py]: more to the point is to push this down to the block manager for execution which handles the mixed dtypes case, this is just a special case of that.

**Review by jreback — APPROVED** [review]
small comments for the future

**Comment by jreback** [comment]
thanks

**Pull request merged**
